File: PSG Officiel/handlers/commands.py
# handlers/commands.py - VERSION AVEC PARAMÈTRE MEMBRE POUR /COLLECTION ET /GIVE
import discord
from discord import app_commands
from discord.ext import commands
from commands.solde import solde_command
from commands.packs import packs_command
from commands.collection import collection_command
from commands.admin import addcoins_command, removecoins_command, setcoins_command
from commands.config import config_command
from commands.give import give_command

def setup_commands(bot: commands.Bot):
    
    # Wrapper pour logger les commandes APRÈS leur exécution (non bloquant)
    async def log_command_wrapper(interaction: discord.Interaction, command_name: str, func, *args, **kwargs):
        """Wrapper qui log l'utilisation de toutes les commandes"""
        from utils.logs import log_command_use
        
        try:
            # ⭐ IMPORTANT: Exécuter la commande EN PREMIER
            await func(interaction, *args, **kwargs)
            
            # ⭐ Logger APRÈS (en arrière-plan, ne bloque pas)
            # Créer une tâche asyncio pour logger sans bloquer
            bot.loop.create_task(log_command_use(interaction, command_name, success=True))
            
        except discord.errors.NotFound:
            # Interaction expirée, ignorer silencieusement
            pass
        except Exception as e:
            # Logger l'erreur
            bot.loop.create_task(log_command_use(interaction, command_name, success=False, error=str(e)))
            # Re-lever l'exception pour que Discord.py la gère
            raise
    
    # ========== COMMANDES MEMBRES ==========
    @bot.tree.command(name="solde", description="Consulte ta solde de PSG Coins")
    async def solde(interaction: discord.Interaction):
        await log_command_wrapper(interaction, "solde", solde_command)
    
    @bot.tree.command(name="packs", description="Voir les packs disponibles et acheter avec des boutons")
    async def packs(interaction: discord.Interaction):
        await log_command_wrapper(interaction, "packs", packs_command)
    
    @bot.tree.command(name="collection", description="Voir ta collection de cartes ou celle d'un autre membre")
    @app_commands.describe(
        membre="(Optionnel) Le membre dont tu veux voir la collection"
    )
    async def collection(interaction: discord.Interaction, membre: discord.Member = None):
        await log_command_wrapper(interaction, "collection", collection_command, membre)
    
    # ========== COMMANDES ADMINISTRATEUR ==========
    @bot.tree.command(name="addcoins", description="[ADMIN] Ajouter des PSG Coins à un membre")
    @app_commands.describe(
        membre="Le membre qui va recevoir les coins",
        montant="Nombre de PSG Coins à ajouter"
    )
    async def addcoins(interaction: discord.Interaction, membre: discord.Member, montant: int):
        await log_command_wrapper(interaction, f"addcoins membre:{membre.name} montant:{montant}", addcoins_command, membre, montant)
    
    @bot.tree.command(name="removecoins", description="[ADMIN] Retirer des PSG Coins à un membre")
    @app_commands.describe(
        membre="Le membre qui va perdre les coins",
        montant="Nombre de PSG Coins à retirer"
    )
    async def removecoins(interaction: discord.Interaction, membre: discord.Member, montant: int):
        await log_command_wrapper(interaction, f"removecoins membre:{membre.name} montant:{montant}", removecoins_command, membre, montant)
    
    @bot.tree.command(name="setcoins", description="[ADMIN] Définir la solde exacte d'un membre")
    @app_commands.describe(
        membre="Le membre dont tu veux modifier la solde",
        montant="Nouvelle solde en PSG Coins"
    )
    async def setcoins(interaction: discord.Interaction, membre: discord.Member, montant: int):
        await log_command_wrapper(interaction, f"setcoins membre:{membre.name} montant:{montant}", setcoins_command, membre, montant)
    
    @bot.tree.command(name="give", description="[ADMIN] Donner une carte à un membre")
    @app_commands.describe(
        carte_id="L'ID de la carte à donner (ex: gk_donnarumma_basic)",
        membre="Le membre qui va recevoir la carte",
        raison="(Optionnel) Raison ou message pour le membre"
    )
    async def give(interaction: discord.Interaction, carte_id: str, membre: discord.Member, raison: str = None):
        await log_command_wrapper(interaction, f"give carte:{carte_id} membre:{membre.name}", give_command, carte_id, membre, raison)
    
    # ========== COMMANDE CONFIGURATION UNIQUE ==========
    @bot.tree.command(name="config", description="[OWNER] Configurer le bot de manière interactive")
    async def config(interaction: discord.Interaction):
        await log_command_wrapper(interaction, "config", config_command)
    
    return bot

File: src/handlers/commands.js
// src/handlers/commands.js - Routeur central des slash commands et interactions boutons/menus
const { logCommandUse } = require('../utils/logs');

const { soldeCommand } = require('../commands/solde');
const { packsCommand, buyPack } = require('../commands/packs');
const { collectionCommand, handleCollectionInteraction } = require('../commands/collection');
const { addCoinsCommand, removeCoinsCommand, setCoinsCommand } = require('../commands/admin');
const { giveCommand } = require('../commands/give');
const { configCommand, handleConfigInteraction } = require('../commands/config');
const { handleMinigameAnswer } = require('../commands/minigame');

function setupCommands(client) {

  client.on('interactionCreate', async (interaction) => {

    // ==================== SLASH COMMANDS ====================
    if (interaction.isChatInputCommand()) {
      const { commandName } = interaction;

      try {
        switch (commandName) {
          case 'solde':
            await soldeCommand(interaction);
            break;

          case 'packs':
            await packsCommand(interaction);
            break;

          case 'collection': {
            const membre = interaction.options.getMember('membre') || null;
            await collectionCommand(interaction, membre);
            break;
          }

          case 'addcoins': {
            const membre = interaction.options.getMember('membre');
            const montant = interaction.options.getInteger('montant');
            await addCoinsCommand(interaction, membre, montant);
            break;
          }

          case 'removecoins': {
            const membre = interaction.options.getMember('membre');
            const montant = interaction.options.getInteger('montant');
            await removeCoinsCommand(interaction, membre, montant);
            break;
          }

          case 'setcoins': {
            const membre = interaction.options.getMember('membre');
            const montant = interaction.options.getInteger('montant');
            await setCoinsCommand(interaction, membre, montant);
            break;
          }

          case 'give': {
            const carteId = interaction.options.getString('carte_id');
            const membre = interaction.options.getMember('membre');
            const raison = interaction.options.getString('raison') || null;
            await giveCommand(interaction, carteId, membre, raison);
            break;
          }

          case 'config':
            await configCommand(interaction);
            break;

          default:
            await interaction.reply({ content: '❌ Commande inconnue.', flags: MessageFlags.Ephemeral });
        }

        // Logger après exécution (non bloquant)
        logCommandUse(interaction, commandName, true).catch(() => {});

      } catch (error) {
        console.error(`❌ Erreur commande /${commandName}:`, error);
        logCommandUse(interaction, commandName, false, error.message).catch(() => {});

        const errMsg = { content: '❌ Une erreur est survenue lors de l\'exécution de cette commande.', flags: MessageFlags.Ephemeral };
        try {
          if (interaction.replied || interaction.deferred) {
            await interaction.followUp(errMsg);
          } else {
            await interaction.reply(errMsg);
          }
        } catch { /* interaction expirée */ }
      }
    }

    // ==================== BOUTONS ====================
    else if (interaction.isButton()) {
      const { customId } = interaction;

      try {
        // Acheter un pack (boutique)
        if (customId.startsWith('buy_pack_')) {
          const parts = customId.split('_');
          // format: buy_pack_{packKey}_{userId}
          const userId = parts[parts.length - 1];
          if (interaction.user.id !== userId) {
            return interaction.reply({ content: '❌ Tu ne peux pas utiliser cette boutique !\n\nUtilise `/packs` pour ouvrir la tienne.', flags: MessageFlags.Ephemeral });
          }
          const packKey = parts.slice(2, parts.length - 1).join('_');
          await buyPack(interaction, packKey);
          return;
        }

        // Mini-jeu
        if (customId.startsWith('minigame_answer_')) {
          await handleMinigameAnswer(interaction);
          return;
        }

        // Collection navigation
        if (customId.startsWith('collection_')) {
          await handleCollectionInteraction(interaction);
          return;
        }

        // Configuration
        if (customId.startsWith('config_')) {
          await handleConfigInteraction(interaction);
          return;
        }

      } catch (error) {
        console.error(`❌ Erreur bouton ${interaction.customId}:`, error);
        try {
          if (!interaction.replied && !interaction.deferred) {
            await interaction.reply({ content: '❌ Une erreur est survenue.', flags: MessageFlags.Ephemeral });
          }
        } catch { /* ok */ }
      }
    }

    // ==================== SELECT MENUS ====================
    else if (interaction.isStringSelectMenu()) {
      const { customId } = interaction;

      try {
        // Collection
        if (customId.startsWith('collection_')) {
          await handleCollectionInteraction(interaction);
          return;
        }

        // Configuration
        if (customId.startsWith('config_')) {
          await handleConfigInteraction(interaction);
          return;
        }

      } catch (error) {
        console.error(`❌ Erreur select menu ${interaction.customId}:`, error);
        try {
          if (!interaction.replied && !interaction.deferred) {
            await interaction.reply({ content: '❌ Une erreur est survenue.', flags: MessageFlags.Ephemeral });
          }
        } catch { /* ok */ }
      }
    }
  });
}

module.exports = { setupCommands };